torch/onnx/symbolic.py

```diff
@@ -693,16 +693,49 @@ def symbolic_fn(g, input, kernel_size, stride, padding, ceil_mode, count_include
 avg_pool3d = _avg_pool('avg_pool3d', _triple)
 
 
-@parse_args('v', 'is')
-def adaptive_avg_pool2d(g, input, output_size):
-    assert output_size == [1, 1], "Only output_size=[1, 1] is supported"
-    return g.op("GlobalAveragePool", input)
+def _adaptive_pool(name, type, tuple_fn, fn=None):
+    @parse_args('v', 'is')
+    def symbolic_fn(g, input, output_size):
+        # _adaptive_pool is supported for cases where output_size is 1 for all dimensions,
+        # by executing a GlobalPool.
+        # It is also supported for cases where the output size is a factor of the input size.
+        # For these cases the stride and kernel size are uniform along all the indices of
+        # the same dimension, which makes it possible to export it to ONNX.
+        # for MaxPool, GlobalMaxPool does not return indices,
+        # so we try using max_poolxd_with_indices, and if it is not possible
+        # (input is not CompleteTensorType or output size not factor of input size)
+        # then we call GlobalAveragePool and return None for the indices
+        if output_size == [1] * len(output_size) and type == "AveragePool":
+            return g.op("GlobalAveragePool", input)
```

Comment:
why not put "GlobalAveragePool" and "GlobalMaxPool" both under this branch?

Reply:
GlobalMaxPool does not return indices information, so we will have to return None for the indices when it is used.
That's why GlobalMaxPool is only called when max_poolxd_with_indices cannot be called (when the input is not CompleteTensorType or the output size not a factor of input size).
Let me know if the comment is not clear and I'll try reformulating :)

Reply:
Yes, I understand that. My point is that you can reorganize the code a bit. 

```suggestion
           if ouptut_size == [1] * len(output_size):
                  if type == "AveragePool":
                        return g.op("GlobalAveragePool", input)
                   if type == "MaxPool":
                         return g.op("GlobalMaxPool", input), None
            if input.type().kind() != "CompleteTensorType":
                 return _unimplemented(name, 'input size not accesible')
```

Reply:
Yes, that would be clearer and easier to read, but I still think we would be missing support to one of the cases.

The difference occurs with the following example:
AdaptiveMaxPool1d((1), **return_indices=True**), with any input shape.
In the proposed code, since ouptut_size == [1] * len(output_size) and type == "MaxPool", then 'GlobalMaxPool' will be called and None will be returned for the indices. 
So exporting the indices will fail. 
But by calling 'max_pool1d_with_indices' instead of 'GlobalMaxPool', we could get the indices output and return it successfully, that's why 'GlobalMaxPool' is not called right ahead.